sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py

```diff
@@ -0,0 +1,224 @@
+
+    zero_idx = np.int64(0)
+    one_incr = np.int32(1)
+    one_decr = np.int32(-1)
+    zero_init = dtype(0.0)
+    max_candidate_id = np.int32(n_samples - 1)
+
+    @dpex.kernel
+    # fmt: off
+    def sample_center_candidates(
+        closest_dist_sq,          # IN             (n_features, n_samples)
+        total_potential,          # IN             (1,)
+        random_state,             # INOUT          (n_local_trials, 2)
+        candidates_id,            # OUT            (n_local_trials,)
+        ):
+    # fmt: on
+        local_trial_idx = dpex.get_global_id(zero_idx)
+        if local_trial_idx >= n_local_trials:
+            return
+        random_value = (rand_uniform_kernel_func(random_state, local_trial_idx)
+                        * total_potential[zero_idx])
+
+        cumulative_potential = zero_init
+        candidate_id = one_decr
+        while (random_value > cumulative_potential) and (candidate_id < max_candidate_id):
+            candidate_id += one_incr
+            cumulative_potential += closest_dist_sq[candidate_id]
+        candidates_id[local_trial_idx] = candidate_id
+
+    global_size = (math.ceil(n_local_trials / work_group_size)) * (work_group_size)
+    return sample_center_candidates[global_size, work_group_size]
+
+
+@lru_cache
+def make_kmeansplusplus_single_step_fixed_window_kernel(
+    n_samples,
+    n_features,
+    n_candidates,
+    preferred_work_group_size_multiple,
+    centroids_window_width_multiplier,
+    centroids_window_height,
+    work_group_size,
+    dtype,
+):
+
+    window_n_centroids = (
+        preferred_work_group_size_multiple * centroids_window_width_multiplier
+    )
+
+    (
+        _initialize_window_of_centroids,
+        _load_window_of_centroids_and_features,
+    ) = _make_initialize_window_kernel_funcs(
+        n_samples,
+        n_features,
+        work_group_size,
+        window_n_centroids,
+        centroids_window_height,
+        dtype,
+    )
+
+    _accumulate_sq_distances = _make_accumulate_sum_of_ops_kernel_func(
+        n_samples,
+        n_features,
+        centroids_window_height,
+        window_n_centroids,
+        ops="squared_diff",
+        dtype=dtype,
+    )
+
+    n_windows_per_feature = math.ceil(n_candidates / window_n_centroids)
+    n_windows_per_centroid = math.ceil(n_features / centroids_window_height)
+
+    centroids_window_shape = (centroids_window_height, (window_n_centroids + 1))
+
+    zero_idx = np.int64(0)
+
+    @dpex.kernel
+    # fmt: off
+    def kmeansplusplus_single_step(
+        X_t,                               # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,                     # IN READ-ONLY   (n_samples,)
+        candidates_ids,                    # IN             (n_candidates,)
+        closest_dist_sq,                   # IN             (n_samples,)
+        euclidean_distances_t,             # OUT            (n_candidates, n_samples)
```

Comment:
+1 for `sq_distances_t`, the naming is inspired from sklearn implementation but it's true that `euclidean_distances_t` make sense in sklearn because the full pairwise distance matrix is fully computed, while it's fused with the `min` step in our kernel.